src/julia-syntax.scm

```diff
@@ -86,35 +86,46 @@
 (define (expand-compare-chain e)
   (car (expand-vector-compare e)))
 
-;; return the appropriate computation for an `end` symbol for indexing
+;; return the appropriate computation for a `begin` or `end` symbol for indexing
 ;; the array `a` in the `n`th index.
 ;; `tuples` are a list of the splatted arguments that precede index `n`
 ;; `last` = is this last index?
-;; returns a call to endof(a) or size(a,n)
+;; returns a call to, e.g., endindex(a) or last(axes(a,n))
 (define (end-val a n tuples last)
   (if (null? tuples)
       (if (and last (= n 1))
-          `(call (top endof) ,a)
-          `(call (top size) ,a ,n))
+          `(call (top endindex) ,a)
+          `(call (top last) (call (top axes) ,a ,n)))
```

Comment:
I like this a lot, but could we incorporate this behavior into `endindex`, too?  That way the parser only lowers `end` to `endindex(A)` or `endindex(A, n)` for the one- and n-arg cases, respectively.